Standards/scs-XXXX-vN-key-manager-standard.md

```diff
@@ -0,0 +1,73 @@
+Key-managers like Barbican provide this functionality on IaaS-Level.
+Using such a key-manger and thus provide to customers the possibility to encrypt user data is not yet implemented everywhere in IaaS deployments.
+A first step towards more security is to encourage CSPs to use this service and provide the feature of using encrypted resources to customers.
+It is also important to take a closer look into the key-manager and to apply aim for an appropiate level of security there.
+The Key-Manager is responsible for storing the keys securely, which can be done by encrypting them with a KEK and storing them in a database and store the KEK either also encrypted in the database or to store it somewhere else.
+Barbican as the OpenStack implementation of a key-manager is relying on a Master-KEK, which encrypts project-specific KEKs, which encrypt the Keys within a project.
+All keys except for the Master-KEK are stored in a database.
+This Master-KEK is stored differently for each plugin and needs to be protected.
+
+## Design Considerations
+
+While discussing what this standard should aim for it was discovered that some CSPs don't use Barbican or another key-manager at all and do not provide the feature to encrypt user data to their customers.
+This should change, but the exact change comes with financial burden, when choosing a plugin in Barbican to store the Master-KEK.
+To minimize the burden and enable more CSPs to step up and provide encryption, this standard will only make recommendations about plugins.
+
+### Options considered
+
+#### _Option 1_
+
+Barbican as the OpenStack key-manager implementation offers different plugins to store the Master-KEK.
+It was considered to only recommend a certain set of plugins, but this may be very prone to change if Barbican adds a new plugin and this standard would need to consider other possible key-managers as well.
+As the SCS only wants to mandate the API that can be abstracted through the Castellan library in OpenStack, integrating any other key-manager implementation is not uncommon.
+So this Option was disregarded.
+
+#### _Option 2_
+
+Looking into the plugins and possible attack vectors one design decision in the plugins is very important: where and how to store the Master-KEK.
+Because the Plugins might use different technologies, but most of them increase the security level by not storing the Master-KEK in plain text on the physical machine Barbican is running on.
+This mechanism as a whole, is something that CSPs should aim to do.
+
+## Standard
+
+To increase the level of security and overall user data encryption CSPs SHOULD implement the key-manager API (e.g. implemented by barbican) with a security level of storing Keys encrypted and storing the KEK in another place than the Keys.
+
+If possible CSPs SHOULD NOT store the Master-KEK in plain-text on the physical host the key-manager is running on.
```

Comment:
I think it would be helpful here to explicitly call out the Barbican plugins which do store keys unencrypted on local disc, so CSPs can avoid these, or do the reverse and at least list some plugins which are deemed "secure" (enough).

Reply:
The problem here is:
the simple crypto plugin does store the masterKEK unencrypted in the Barbican config. So we can name it as a negative example. But I do not think we should deem this plugin as completely insecure.
E.g. it can be managed to let Barbican run in enclaves like SGX, to protect the config.
And having Barbican with simpley crypto is still better than not having any option to encrypt user data at all - so I am a bit hesitant to give a "not do this"-example.

We could add a sentence stating, that using HSMs would be one option to fulfill this, but this could give CSPs the impression, that we will put a high financial burden on them.

I am skeptical of both options and would like to have more input from other reviewers on this

Reply:
I would add another sentence like: "Barbican supports deploying out-of-tree drivers what enables operators to satisfy their specific needs." With this sentence I would like to give a signal to the CSPs that they have quite lot possibilities in their hands to offer better security to their customers.

Reply:
@gtema this is more of an implementation note and not part of the standard. Should I still add this somewhere in this document or just reference the implementation notes somewhere?

Maybe an:
```
:::tip

Barbican supports deploying out-of-tree drivers what enables operators to satisfy their specific needs.

:::
```
should be added?

Reply:
yupp, adding a tip sound good (I agree that we can place it in the impl. part). I just want to ensure, that CSPs are not frightened by the standard and know that it is possible to have some custom stuff running what would satisfy the standard.

Reply:
from my pov this conversation is resolved, @gtema ?

Reply:
You need barbican anyway alone as a requirement for octavia if you do ssl offloading with amphoras too. Also other Keymanager appliances plugin into "barbican" as drivers/plugins